File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/smartcollection/SmartCollectionRepository.kt
/*
 * SmartCollectionRepository.kt
 *
 * v1 rule evaluator + seed for the Workspace Smart Collections
 * surface (Phase C.3). Matches captures against a [RuleClause]
 * list in-memory: at v1 corpus sizes (low thousands) the join-free
 * Kotlin filter is sub-millisecond and avoids a SQL-builder
 * surface area we don't need yet. Phase D can swap for a typed
 * @RawQuery if the corpus outgrows it.
 *
 * Seeded rows are flagged `is_seeded = 1` so later releases can
 * silently rev the rule_json without overwriting user-edited
 * collections (per brief §2 entity notes).
 */

package app.quickink.mobile.data.smartcollection

import app.quickink.mobile.data.capture.CaptureDao
import app.quickink.mobile.data.capture.CaptureEntity
import app.quickink.mobile.data.capturetag.CaptureTagDao
import app.quickink.mobile.data.tag.TagDao
import app.quickink.mobile.data.tag.TagRepository
import app.releaf.mobile.data.common.IsoClock
import app.releaf.mobile.data.common.Uuidv7
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flowOf
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.IsoFields

class SmartCollectionRepository(
    private val smartCollectionDao: SmartCollectionDao,
    private val captureDao: CaptureDao? = null,
    private val captureTagDao: CaptureTagDao? = null,
    private val tagDao: TagDao? = null,
) {

    fun observeActive(userId: String): Flow<List<SmartCollectionEntity>> =
        smartCollectionDao.observeActive(userId)

    suspend fun findById(id: String): SmartCollectionEntity? =
        smartCollectionDao.findById(id)

    /**
     * Stream the captures matching [collection]'s rule. Composes
     * the captures-by-user Flow with the per-capture tag-id Flow,
     * filtering in-memory. Re-emits on any change to either side.
     * Empty list when DAOs aren't wired or the rule is malformed.
     */
    fun observeMatchingCaptures(
        userId: String,
        collection: SmartCollectionEntity,
    ): Flow<List<CaptureEntity>> {
        val capDao = captureDao ?: return flowOf(emptyList())
        val joinDao = captureTagDao ?: return flowOf(emptyList())
        val rule = SmartCollectionRule.decode(collection.ruleJson)
        if (rule.isEmpty()) return flowOf(emptyList())

        return combine(
            capDao.observeActive(userId),
            joinDao.observeTagCounts(userId),
        ) { captures, _ ->
            captures.filter { matchesAllClauses(it, rule, userId, joinDao) }
        }
    }

    /**
     * Synchronous best-effort match for static contexts (e.g.
     * worker code that already has the captures in hand). Reads
     * tag attachments per row from the DAO; not Flow-based.
     */
    private fun matchesAllClauses(
        capture: CaptureEntity,
        clauses: List<RuleClause>,
        userId: String,
        joinDao: CaptureTagDao,
    ): Boolean {
        for (clause in clauses) {
            if (!matchesClause(capture, clause, userId, joinDao)) return false
        }
        return true
    }

    private fun matchesClause(
        capture: CaptureEntity,
        clause: RuleClause,
        userId: String,
        joinDao: CaptureTagDao,
    ): Boolean = when (clause) {
        is RuleClause.FolderIs ->
            capture.folderId == clause.folderId

        is RuleClause.TagIs ->
            // suspend lookup costs a coroutine context — read
            // via runBlocking is OK here because Flow combine
            // dispatches on Default. Bounded by the join row
            // count per capture (typically <12 per the soft cap).
            kotlinx.coroutines.runBlocking {
                joinDao.listTagIdsForCapture(capture.id).contains(clause.tagId)
            }

        is RuleClause.TagIsNot ->
            kotlinx.coroutines.runBlocking {
                !joinDao.listTagIdsForCapture(capture.id).contains(clause.tagId)
            }

        is RuleClause.DateRange -> {
            val timestamp = when (clause.field) {
                "created_at"     -> capture.createdAt
                "last_opened_at" -> capture.lastOpenedAt ?: return false
                else             -> return false
            }
            inDatePreset(timestamp, clause.preset)
        }

        is RuleClause.SourceIs ->
            capture.source == clause.value

        // OCR-derived signals — placeholder until Phase E. Always
        // false so the rule matches nothing rather than everything.
        is RuleClause.HasHandwriting,
        is RuleClause.HasSignature,
        is RuleClause.HasOcrText -> false
    }

    /**
     * Resolves a preset like "this_month" against the device's
     * local time zone. Returns true when the ISO timestamp falls
     * inside the bucket.
     */
    private fun inDatePreset(isoTimestamp: String, preset: String): Boolean {
        val zone = ZoneId.systemDefault()
        val ts = runCatching { java.time.Instant.parse(isoTimestamp) }.getOrNull() ?: return false
        val tsDate = ts.atZone(zone).toLocalDate()
        val today = LocalDate.now(zone)

        return when (preset) {
            "this_week" -> {
                val weekFields = java.time.temporal.WeekFields.ISO
                tsDate.get(weekFields.weekOfWeekBasedYear()) == today.get(weekFields.weekOfWeekBasedYear()) &&
                    tsDate.year == today.year
            }
            "this_month" ->
                tsDate.year == today.year && tsDate.month == today.month
            "last_30_days" ->
                !tsDate.isBefore(today.minusDays(30))
            "this_quarter" ->
                tsDate.year == today.year &&
                    tsDate.get(IsoFields.QUARTER_OF_YEAR) == today.get(IsoFields.QUARTER_OF_YEAR)
            else -> false
        }
    }

    // ────────────────────────────────────────────────────────────
    // Seeding (Phase A.3-equivalent for smart collections)
    // ────────────────────────────────────────────────────────────

    /**
     * Idempotent seed of the Workspace v1 default smart
     * collections — currently just "Needs review", which depends
     * on the `#needs-review` tag seeded in [TagRepository.DEFAULT_SEED].
     *
     * "Invoices this month" and "Contains signatures" from the
     * design brief are deferred:
     *   - Invoices this month: requires an "Invoices" folder seed.
     *     Folder-name choices are user territory; we don't seed.
     *   - Contains signatures: requires has_signature, which is
     *     Phase E OCR-derived metadata.
     */
    suspend fun seedDefaultsIfNeeded(userId: String) {
        val tags = tagDao ?: return
        val existing = smartCollectionDao.listSeeded(userId)
        if (existing.any { it.name == SEED_NEEDS_REVIEW }) return

        val needsReviewTag = tags.findByName(userId, "needs-review") ?: return
        val now = IsoClock.nowIso()
        smartCollectionDao.insert(
            SmartCollectionEntity(
                id        = Uuidv7.generate(),
                userId    = userId,
                name      = SEED_NEEDS_REVIEW,
                icon      = "ti-eye",
                color     = "#E8AE17",
                ruleJson  = SmartCollectionRule.encode(
                    listOf(RuleClause.TagIs(needsReviewTag.id)),
                ),
                position  = 0,
                isSeeded  = true,
                createdAt = now,
                updatedAt = now,
                dirty     = true,
                deletedAt = null,
            ),
        )
    }

    companion object {
        const val SEED_NEEDS_REVIEW = "Needs review"
    }
}

File: apps/quickink/ios/QuickInk/Data/SmartCollectionRepository.swift
/*
 * SmartCollectionRepository.swift
 *
 * iOS rule evaluator + seed mirror of Android
 * `SmartCollectionRepository.kt`. In-memory filter: corpus-size-
 * appropriate at v1, with a clear path to a typed raw-SQL builder
 * once the workspace grows.
 *
 * Seeded "Today" is date-only; "Needs review" depends on the
 * `#needs-review` tag landing via TagRepository.defaultSeed first.
 */

import Foundation
import Combine
import GRDB
import ReleafCoreData

public final class SmartCollectionRepository: @unchecked Sendable {

    public static let seedToday = "Today"
    public static let seedNeedsReview = "Needs review"
    private static let seedBlue = "#3A78AE"

    private let dbQueue: DatabaseQueue
    private let tagRepository: TagRepository

    public init(database: QuickInkDatabase = .shared) {
        self.dbQueue = database.dbQueue
        self.tagRepository = TagRepository(database: database)
    }

    // MARK: - Reads

    public func observeActive(userId: String) -> AnyPublisher<[SmartCollectionEntity], Error> {
        ValueObservation.tracking { [userId] db in
            try SmartCollectionEntity
                .filter(Column("user_id") == userId)
                .filter(Column("deleted_at") == nil)
                .order(Column("position").asc, Column("name").asc)
                .fetchAll(db)
        }
        .publisher(in: dbQueue)
        .eraseToAnyPublisher()
    }

    public func findById(_ id: String) async throws -> SmartCollectionEntity? {
        try await dbQueue.read { db in
            try SmartCollectionEntity.filter(Column("id") == id).fetchOne(db)
        }
    }

    public func observeMatchingCaptures(
        userId: String,
        collection: SmartCollectionEntity
    ) -> AnyPublisher<[CaptureSummary], Error> {
        let clauses = SmartCollectionRule.decode(collection.ruleJson)
        guard !clauses.isEmpty else {
            return Just([CaptureSummary]())
                .setFailureType(to: Error.self)
                .eraseToAnyPublisher()
        }
        return ValueObservation.tracking { [userId, clauses] db -> [CaptureSummary] in
            let captures = try CaptureSummary.fetchAll(db, sql: """
                SELECT id, title, preview_uri, pdf_uri, page_count, created_at, source,
                       latitude, longitude, locality, sub_locality, address,
                       folder_id, last_opened_at, last_opened_page, last_opened_device
                FROM captures
                WHERE user_id = ? AND deleted_at IS NULL
                ORDER BY created_at DESC
                """, arguments: [userId])
            return captures.filter { cap in
                clauses.allSatisfy { clause in
                    Self.matches(clause: clause, capture: cap, db: db)
                }
            }
        }
        .publisher(in: dbQueue)
        .eraseToAnyPublisher()
    }

    private static func matches(clause: RuleClause, capture: CaptureSummary, db: Database) -> Bool {
        switch clause {
        case .folderIs(let id):
            return capture.folderId == id
        case .tagIs(let tagId):
            let n = (try? Int.fetchOne(db, sql: """
                SELECT COUNT(*) FROM capture_tags
                WHERE capture_id = ? AND tag_id = ? AND deleted_at IS NULL
                """, arguments: [capture.id, tagId])) ?? 0
            return n > 0
        case .tagIsNot(let tagId):
            let n = (try? Int.fetchOne(db, sql: """
                SELECT COUNT(*) FROM capture_tags
                WHERE capture_id = ? AND tag_id = ? AND deleted_at IS NULL
                """, arguments: [capture.id, tagId])) ?? 0
            return n == 0
        case .dateRange(let field, let preset):
            let ts: String
            switch field {
            case "created_at":     ts = capture.createdAt
            case "last_opened_at": guard let t = capture.lastOpenedAt else { return false }; ts = t
            default: return false
            }
            return inDatePreset(timestamp: ts, preset: preset)
        case .sourceIs(let value):
            return capture.source == value
        case .hasHandwriting, .hasSignature, .hasOcrText:
            // OCR-derived placeholders; flagged on but column not
            // emitted yet — return false until Phase E.
            return false
        }
    }

    private static func inDatePreset(timestamp: String, preset: String) -> Bool {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = formatter.date(from: timestamp) ??
                ISO8601DateFormatter().date(from: timestamp) else {
            return false
        }
        let cal = Calendar.current
        let now = Date()
        switch preset {
        case "today":
            return cal.isDateInToday(date)
        case "yesterday":
            return cal.isDateInYesterday(date)
        case "this_week":
            return cal.isDate(date, equalTo: now, toGranularity: .weekOfYear)
        case "this_month":
            return cal.isDate(date, equalTo: now, toGranularity: .month)
        case "last_30_days":
            guard let cutoff = cal.date(byAdding: .day, value: -30, to: now) else { return false }
            return date >= cutoff
        case "this_quarter":
            let q = (cal.component(.month, from: date) - 1) / 3
            let qNow = (cal.component(.month, from: now) - 1) / 3
            return q == qNow && cal.isDate(date, equalTo: now, toGranularity: .year)
        default:
            return false
        }
    }

    // MARK: - Seed

    public func seedDefaultsIfNeeded(userId: String) async throws {
        let existing = try await dbQueue.read { db in
            try SmartCollectionEntity
                .filter(Column("user_id") == userId)
                .filter(Column("is_seeded") == true)
                .filter(Column("deleted_at") == nil)
                .fetchAll(db)
        }
        let existingNames = Set(existing.map(\.name))
        let now = IsoClock.nowIso()

        if !existingNames.contains(Self.seedToday) {
            let row = SmartCollectionEntity(
                id:        Uuidv7.generate(),
                userId:    userId,
                name:      Self.seedToday,
                icon:      "star",
                color:     Self.seedBlue,
                ruleJson:  SmartCollectionRule.encode([.dateRange(field: "created_at", preset: "today")]),
                position:  0,
                isSeeded:  true,
                createdAt: now,
                updatedAt: now,
                dirty:     true,
            )
            try await dbQueue.write { db in
                do { try row.insert(db) }
                catch let e as DatabaseError where e.resultCode == .SQLITE_CONSTRAINT {
                    // Race lost — another seed pass won.
                }
            }
        }

        if var needsReview = existing.first(where: { $0.name == Self.seedNeedsReview }),
           needsReview.position != 1 {
            needsReview.position = 1
            needsReview.updatedAt = now
            needsReview.dirty = true
            try await dbQueue.write { db in
                try needsReview.update(db)
            }
        }

        if existingNames.contains(Self.seedNeedsReview) {
            return
        }
        guard let needsReviewTag = try await tagRepository.findByName(userId: userId, name: "needs-review") else {
            return
        }
        let row = SmartCollectionEntity(
            id:        Uuidv7.generate(),
            userId:    userId,
            name:      Self.seedNeedsReview,
            icon:      "eye",
            color:     "#E8AE17",
            ruleJson:  SmartCollectionRule.encode([.tagIs(tagId: needsReviewTag.id)]),
            position:  1,
            isSeeded:  true,
            createdAt: now,
            updatedAt: now,
            dirty:     true,
        )
        try await dbQueue.write { db in
            do { try row.insert(db) }
            catch let e as DatabaseError where e.resultCode == .SQLITE_CONSTRAINT {
                // Race lost — another seed pass won.
            }
        }
    }
}
